mail/local/mailbox.go

```diff
@@ -0,0 +1,256 @@
+package local
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	tc "github.com/textileio/go-threads/api/client"
+	"github.com/textileio/go-threads/core/db"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/textile/api/common"
+	"github.com/textileio/textile/api/users/client"
+	"github.com/textileio/textile/cmd"
+	"github.com/textileio/textile/mail"
+)
+
+// Mailbox is a local-first messaging library built on ThreadDB and IPFS.
+type Mailbox struct {
+	cwd     string
+	conf    *cmd.Config
+	clients *cmd.Clients
+	id      thread.Identity
+	token   thread.Token
+}
+
+// Identity returns the mailbox's identity.
+func (m *Mailbox) Identity() thread.Identity {
+	return m.id
+}
+
+// SendMessage sends the message body to a recipient.
+func (m *Mailbox) SendMessage(ctx context.Context, to thread.PubKey, body []byte) (msg client.Message, err error) {
+	ctx, err = m.context(ctx)
+	if err != nil {
+		return
+	}
+	return m.clients.Users.SendMessage(ctx, m.id, to, body)
+}
+
+// ListInboxMessages lists messages from the inbox.
+// Use options to paginate with seek and limit,
+// and filter by read status.
+func (m *Mailbox) ListInboxMessages(ctx context.Context, opts ...client.ListOption) ([]client.Message, error) {
+	ctx, err := m.context(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return m.clients.Users.ListInboxMessages(ctx, opts...)
+}
+
+// ListSentMessages lists messages from the sentbox.
+// Use options to paginate with seek and limit.
+func (m *Mailbox) ListSentMessages(ctx context.Context, opts ...client.ListOption) ([]client.Message, error) {
+	ctx, err := m.context(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return m.clients.Users.ListSentMessages(ctx, opts...)
+}
+
+const reconnectInterval = time.Second * 5
+
+// MailboxEvent describes an event that occurred in a mailbox.
+type MailboxEvent struct {
+	// Type of event.
+	Type MailboxEventType
+	// Message identifier.
+	MessageID db.InstanceID
+	// Message will contain the full message unless this is a delete event.
+	Message client.Message
+}
+
+// MailboxEventType is the type of mailbox event.
+type MailboxEventType int
+
+const (
+	// NewMessage indicates the mailbox has a new message.
+	NewMessage MailboxEventType = iota
+	// MessageRead indicates a message was read in the mailbox.
+	MessageRead
+	// MessageDeleted indicates a message was deleted from the mailbox.
+	MessageDeleted
+)
+
+// WatchInbox watches the inbox for new mailbox events.
+// If offline is true, this will keep watching during network interruptions.
+// Returns a channel of watch connectivity states.
+// Cancel context to stop watching.
+func (m *Mailbox) WatchInbox(ctx context.Context, mevents chan<- MailboxEvent, offline bool) (<-chan cmd.WatchState, error) {
```

Comment:
You can watch your inbox (or sentbox below) for new messages, read messages, and deleted messages.